File: js/t3/tests/oracle.js
// oracle.js — an independent, literal interpreter of T3.v. Reuses T2's oracle
// (`t2/tests/oracle.js`) for the embedded s2 state (exactly as T3.v embeds
// T2.State), and hand-rolls HoldPiece's NewPieceState-equivalent independently
// of `t1/model.js`/`model.js` — a bug shared between the translation and the
// helper it reuses (T2Eng.T1Eng.newPieceState) would still surface as a
// mismatch here.

import { oracle2 } from '../../t2/tests/oracle.js';

// spec: NewPieceState p_new s — independent hand-roll (see file header)
function newPieceState(pNew, s1, params) {
  return {
    mg: s1.mg,
    p: pNew,
    py: params.InitialY(pNew),
    px: params.InitialX(pNew),
    pr: 0,
    gameover: s1.gameover,
    clearedLines: s1.clearedLines,
  };
}

function initState3(p, params) {
  return { s2: oracle2.initState(p, params), hold: null, swapped: false };
}

function movePiece3(dy, dx, s, params) {
  const s2p = oracle2.movePiece(dy, dx, s.s2, params);
  return s2p ? { ...s, s2: s2p } : null;
}

function rotatePiece3(cw, s, params) {
  const s2p = oracle2.rotatePiece(cw, s.s2, params);
  return s2p ? { ...s, s2: s2p } : null;
}

function fixPiece3(pNew, s, params) {
  const s2p = oracle2.fixPiece(pNew, s.s2, params);
  if (!s2p) return null;
  return { s2: s2p, hold: s.hold, swapped: false }; // req-hold-limit
}

function fallStep3(pNew, s, params) {
  return movePiece3(-1, 0, s, params) ?? fixPiece3(pNew, s, params);
}

function holdPiece3(pNew, s, params) {
  const gameover = s.s2.s1.gameover;
  if (!(!gameover && !s.swapped)) return null; // real guard (req-hold-limit)
  const p2 = (s.hold !== null) ? s.hold : pNew; // req-hold-swap / req-hold-empty
  const oldP = s.s2.s1.p;                        // read BEFORE overwritten
  const newS1 = newPieceState(p2, s.s2.s1, params);
  return {
    s2: { ...s.s2, s1: newS1 }, // score/level/combo/perfectClear/totalClearedLines untouched
    hold: oldP,                  // req-hold
    swapped: true,                // req-hold-limit
  };
}

function snapshotOf3(s) {
  return {
    ...oracle2.snapshotOf(s.s2),
    hold: s.hold,
    swapped: s.swapped,
  };
}

function states3Equal(a, b) {
  return JSON.stringify(snapshotOf3(a)) === JSON.stringify(snapshotOf3(b));
}

export const oracle3 = {
  makeParams: oracle2.makeParams,
  initState3,
  movePiece3,
  rotatePiece3,
  fixPiece3,
  fallStep3,
  holdPiece3,
  snapshotOf3,
  states3Equal,
};

File: rs/t3/tests/oracle.rs
//! spec: tests/oracle.rs — §9, `implementation.md`.
//!
//! A from-scratch reimplementation of `hold`/`swapped` and `HoldPiece`'s
//! field-update, independent of `t3::model`'s `hold_piece` under test.
//! Hand-rolls the update rather than calling `t1::model::new_piece_state`,
//! so a bug shared between the translation and a reused helper would still
//! surface as a mismatch here (t2's oracle takes the same approach with its
//! saturating-add helper).
//!
//! R-TestScope: the `s2` half of every step is not re-verified here — `t2`'s
//! and `t1`'s own oracles already check it exhaustively, one layer down.
//! R-HoldOrdering: the old current piece is read before a hold mutates it,
//! mirroring what `t3::model::Machine::hold_piece` itself must do.

#[path = "test_instance.rs"]
mod fixture;

use fixture::{Prng, TestInstance};
use t1::model::Params;
use t3::instance::Tetris;
use t3::model::Machine;

// ── Independent hold/swap bookkeeping (spec: HoldPiece, UnchangedT3Part) ──

struct RocqState3<Piece> {
    hold: Option<Piece>,
    swapped: bool,
}

fn rocq_init3<Piece>() -> RocqState3<Piece> {
    RocqState3 {
        hold: None,
        swapped: false,
    }
}

/// spec: `T1.NewPieceState p2 (s1 s2_)`, hand-rolled independently of
/// `t1::model::new_piece_state` — returns the four fields it would set.
fn rocq_new_piece_state<P: Params>(p2: P::Piece) -> (P::Piece, i64, i64, u8) {
    (p2, P::initial_y(p2), P::initial_x(p2), 0)
}

/// spec: `HoldPiece`'s `hold`/`swapped` field-update (R-HoldOrdering: `old_p`
/// must be read before mutation, §4-T3e). `_s` is unused — both fields are
/// written unconditionally once the guard holds — kept as a parameter only
/// so the call site reads as "transition from state s".
fn rocq_hold3<Piece>(_s: &RocqState3<Piece>, old_p: Piece) -> RocqState3<Piece> {
    RocqState3 {
        hold: Some(old_p), // req-hold
        swapped: true,     // req-hold-limit
    }
}

fn equivalent3<P: Params>(m: &Machine<P>, s: &RocqState3<P::Piece>) -> bool {
    m.hold == s.hold && m.swapped == s.swapped
}

// ── Random-trace differential check ───────────────────────────────────────

fn run_differential<P: Params>(seed: u64, steps: usize) {
    let mut rng = Prng::new(seed);
    let pieces = P::piece_all();
    let start = rng.choose(pieces);

    let mut m = Machine::<P>::new(start, || rng.choose(pieces));
    let mut s = rocq_init3::<P::Piece>();
    assert!(equivalent3(&m, &s), "diverged at Init");

    for step in 1..=steps {
        let cmd = rng.next_below(8) as u8;
        let piece = rng.choose(pieces);

        match cmd {
            0 => {
                m.move_piece(0, -1);
                s = RocqState3 {
                    hold: s.hold,
                    swapped: s.swapped,
                }; // UnchangedT3Part
            }
            1 => {
                m.move_piece(0, 1);
                s = RocqState3 {
                    hold: s.hold,
                    swapped: s.swapped,
                };
            }
            2 => {
                m.move_piece(-1, 0);
                s = RocqState3 {
                    hold: s.hold,
                    swapped: s.swapped,
                };
            }
            3 => {
                m.rotate_piece(true);
                s = RocqState3 {
                    hold: s.hold,
                    swapped: s.swapped,
                };
            }
            4 => {
                m.rotate_piece(false);
                s = RocqState3 {
                    hold: s.hold,
                    swapped: s.swapped,
                };
            }
            5 => {
                if m.fix_piece(piece) {
                    s = RocqState3 {
                        hold: s.hold,
                        swapped: false,
                    }; // req-hold-limit
                }
            }
            6 => {
                // spec: FallStep — branch observed directly from move_piece's return value.
                if !m.move_piece(-1, 0)
                    && m.fix_piece(piece) {
                        s = RocqState3 {
                            hold: s.hold,
                            swapped: false,
                        };
                    }
            }
            _ => {
                let guard_holds = !m.s2.s1.gameover && !m.swapped;
                if guard_holds {
                    let old_p = m.s2.s1.p; // R-HoldOrdering: read before mutation
                    let (expected_p, expected_py, expected_px, expected_pr) =
                        rocq_new_piece_state::<P>(s.hold.unwrap_or(piece));
                    let fired = m.hold_piece(piece);
                    assert!(
                        fired,
                        "guard predicted to hold but hold_piece returned false at step {step}"
                    );
                    assert_eq!(
                        m.s2.s1.p, expected_p,
                        "p mismatch after hold at step {step}"
                    );
                    assert_eq!(
                        m.s2.s1.py, expected_py,
                        "py mismatch after hold at step {step}"
                    );
                    assert_eq!(
                        m.s2.s1.px, expected_px,
                        "px mismatch after hold at step {step}"
                    );
                    assert_eq!(
                        m.s2.s1.pr, expected_pr,
                        "pr mismatch after hold at step {step}"
                    );
                    s = rocq_hold3(&s, old_p);
                } else {
                    let fired = m.hold_piece(piece);
                    assert!(
                        !fired,
                        "guard predicted to fail but hold_piece returned true at step {step}"
                    );
                    // s unchanged (stutter)
                }
            }
        }
        assert!(equivalent3(&m, &s), "hold/swapped diverged at step {step}");
    }
}

#[test]
fn oracle_matches_random_trace_on_test_instance() {
    run_differential::<TestInstance>(0x0ff1_ce0d_d1ce_5eed, 5_000);
}

#[test]
fn oracle_matches_random_trace_on_real_instance() {
    run_differential::<Tetris>(0xfeed_face_f00d_cafe, 1_000);
}
